sympy/core/expr.py

```diff
@@ -329,98 +329,80 @@ def __complex__(self):
         re, im = result.as_real_imag()
         return complex(float(re), float(im))
 
-    def __ge__(self, other):
-        from sympy import GreaterThan
+    def _cmp(self, other, op, cls):
+        assert op in ("<", ">", "<=", ">=")
         try:
             other = _sympify(other)
         except SympifyError:
-            raise TypeError("Invalid comparison %s >= %s" % (self, other))
+            raise TypeError("Invalid comparison %s %s %s" % (self, op, other))
         for me in (self, other):
             if me.is_extended_real is False:
                 raise TypeError("Invalid comparison of non-real %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
+
         n2 = _n2(self, other)
         if n2 is not None:
-            return _sympify(n2 >= 0)
+            # use float comparison for infinity.
+            # otherwise get stuck in infinite recursion
+            if n2 in (S.Infinity, S.NegativeInfinity):
+                n2 = float(n2)
+            if op == "<":
+                return _sympify(n2 < 0)
+            elif op == ">":
+                return _sympify(n2 > 0)
+            elif op == "<=":
+                return _sympify(n2 <= 0)
+            else: # >=
+                return _sympify(n2 >= 0)
+
         if self.is_extended_real and other.is_extended_real:
-            if (self.is_infinite and self.is_extended_positive) \
-                    or (other.is_infinite and other.is_extended_negative):
-                return S.true
-            nneg = (self - other).is_extended_nonnegative
-            if nneg is not None:
-                return sympify(nneg)
-        return GreaterThan(self, other, evaluate=False)
+            if op in ("<=", ">") \
+                and ((self.is_infinite and self.is_extended_negative) \
+                        or (other.is_infinite and other.is_extended_positive)):
+                return S.true if op == "<=" else S.false
+            if op in ("<", ">=") \
+                and ((self.is_infinite and self.is_extended_positive) \
+                        or (other.is_infinite and other.is_extended_negative)):
+                return S.true if op == ">=" else S.false
+            diff = self - other
+            if diff is not S.NaN:
+                if op == "<":
+                    test = sympify(diff.is_extended_negative)
+                elif op == ">":
+                    test = sympify(diff.is_extended_positive)
+                else: # <= or >=
+                    if diff.is_zero:
+                        test = S.true
+                    elif op == "<=":
+                        test = sympify(diff.is_extended_nonpositive)
```

Comment:
We don't need to test is_zero separately because if is_zero is True then extended_nonpositive will be True as well.

Reply:
So this if elif can be flattened

Reply:
Okay, I will commit this and push. I also made the small test change mentioned above